invenio_search/ext.py

```diff
@@ -246,48 +247,128 @@ def active_aliases(self):
-            # Iterate over aliases:
+        def _build(tree_or_filename, alias=None):
+            """Create indices and aliases by walking DFS.
+
+            Yields the result from create_index() and boolean is_index.
+            """
             for name, value in tree_or_filename.items():
                 if isinstance(value, dict):
-                    for result in _create(value, alias=name):
+                    for result in _build(value, alias=name):
                         yield result
                 else:
-                    with open(value, 'r') as body:
-                        yield name, self.client.indices.create(
-                            index=name,
-                            body=json.load(body),
+                    index_result, alias_result = \
+                        self.create_index(
+                            name,
                             ignore=ignore,
+                            dry_run=True
                         )
-
+                    if index_result[0]:
+                        created_indices[name] = index_result[0]
+                        yield index_result, name, True
+                    if alias_result[0]:
+                        data = dict(
+                            index=index_result[0],
+                            name=alias_result[0]
+                        )
+                        yield alias_result, data, False
             if alias:
-                yield alias, self.client.indices.put_alias(
-                    index=list(_get_indices(tree_or_filename)),
-                    name=alias,
-                    ignore=ignore,
+                alias_indices = self._get_indices(tree_or_filename)
+                alias_indices = [created_indices[i] for i in alias_indices]
+                alias_name = build_index_name(alias, suffix='', app=self.app)
+                data = dict(
+                    index=alias_indices,
+                    name=alias_name
                 )
+                yield (alias_name, None), data, False
 
-        for result in _create(self.active_aliases):
-            yield result
+        results = list(_build(self.active_aliases))
+
+        for (full_name, _), name, is_index in results:
+            if current_search_client.indices.exists(full_name):
+                raise Exception(
+                    'index or alias with name "{}" already exists'.format(
+                        full_name)
+                )
+
+        # Create indices and aliases
+        create_results = []
+        for res, data, is_index in results:
+            full_name = res[0]
+            create_results.append(res)
+            if is_index:
+                self.create_index(data, create_alias=False)
+            else:
+                self.client.indices.put_alias(
+                    index=data['index'],
+                    name=data['name'],
+                    ignore=ignore,
+                )
+        return create_results
```

Comment:
It might be worth to keep this function as a generator, since the `invenio index init` command expects this in order to render its progress bar